### PR TITLE
BUG: acorr_breusch_godfrey, problem with regressions without const or exogs

### LINKED ISSUE #6642: BUG: acorr_breusch_godfrey error in regressions without const or exogs
In ARMA models without constant (or ARMAX without constant and without exogs) the acorr_breusch_godfrey method don't check if res.model.exog is None and try to use exogs returning a ValueError.

#### Code Sample, a copy-pastable example if possible

```python
from statsmodels.tsa.arima_model import ARMA
import pandas as pd
from statsmodels.datasets import sunspots

data = sunspots.load_pandas().data['SUNACTIVITY']

res = ARMA(data, (1,0)).fit(disp=False, trend = 'nc')

from statsmodels.stats.diagnostic import acorr_breusch_godfrey

acorr_breusch_godfrey(res, nlags=1)
```

#### Expected Output

N/A

#### Output of ``import statsmodels.api as sm; sm.show_versions()``

0.11.0

### PR DESCRIPTION
**Pull request by ricardozago**
In ARMA models without constant (or ARMAX without constant and without exogs) the acorr_breusch_godfrey method don't check if res.model.exog is None.

If it's None then where is a ValueError. I add a check before use exog, if is None, then we use only the lags terms.

- [X] closes #6642
- [X] tests added / passed. 
- [X] code/documentation is well formatted.  
- [X] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy-1.15.1/dev/gitwash/development_workflow.html#writing-the-commit-message).

### CODE REVIEW

**Comment by josef-pkt** [comment]
In principle the change looks good to me
Unit test should `assert` something, at least regression tests (i.e. current numbers)

However, I'm not sure it applies to ARMA without other regressors in the auxiliary regression.
If there is a trend in the arma, then that should be included, (I'm not sure trend is part of exog in the various tsa models).
But I don't know whether endogenous regressors (lagged endog from AR(MA) model need to be included (my guess would be that they do need to be included).

I never looked at Breusch-Godfrey for other models than regression models (OLS, for which it was written).

quick google search:
https://stats.stackexchange.com/questions/148004/testing-for-autocorrelation-ljung-box-versus-breusch-godfrey
The answer says Breusch-Godfrey applies to ARMA, (but is not explicit in whether lagged endog need to be included).
According to that answer, there is a problem with Ljung-Box with ARMA residuals, but I think that is the standard autocorrelation test in use.

https://www.stata.com/meeting/new-orleans13/abstracts/materials/nola13-baum.pdf
p 21. He estimates and AR using OLS with lagged endoq. In that case BG would add those lagged endog as regressors.

Maybe @ChadFulton and @bashtage know more about how this test can be applied with ARMA

**Comment by josef-pkt** [comment]
For example estimating an ARMA(p,0), i.e. AR(p) and comparing with OLS with p-lagged endog as exog.
BG in those two should produce the same results in larger samples where initial observations are negligible, or when ARMA is estimated without full MLE.

My guess is that just running acorr_breusch_godfrey on the AR/ARMA residuals will not match the test after OLS.
BG might be the same (asymptotically equivalent) to ljung-box if there are no exog

**Comment by josef-pkt** [comment]
stackoverflow question without answer asking for BG after ARIMA
https://stackoverflow.com/questions/58005571/breusch-godfrey-test-on-residuals-from-arima-models
R packages seem to raise and refuse to compute without exog.

My guess is that we could get at least BG for an AR with lagged endog in the auxiliary regression as valid method, given the analogy to OLS. But I'm not sure about an MA term in an ARMA.

I also guess BG after ARX/ARMAX with only the X is not valid, or needs stronger assumptions.

Several references on stackoverflow are to Hayashi. But I'm not familiar with this tsa literature.

**Comment by josef-pkt** [comment]
I had a quick look at Hayashi, section 2.10 on serial correlation:
The setup is OLS with predetermined variables in `x`, so includes AR. Does not say anything about AR/ARMA.
Hayashi has a modified ljung-box (from my quick look, I cannot tell if that is the same as we have) that requires only weak exogeneity (which I guess contradicts the stackoverflow answer that ljung-box doesn't apply at all to time series models because it requires strict exogeneity.)
He also states that the modified ljung-box is equivalent to BG, and that BG is a somputationally convenient way, to get an equivalent test. (e.g. excercise 11 p. 173)

(So that's still only the OLS and non-tsa model context that I had when I wrote BG initially)

**Comment by josef-pkt** [comment]
aside: current implementation uses centered rsquared
`lm = nobs * resols.rsquared`
(It looks like all lm test in `diagnostic.py` use centered rsquared.)

This is identical to uncentered rsquared in OLS, or in general when mean(resid)=0.
Hayashi p.148
This does not extend directly to models where mean(resid) can be different from zero.
(e.g. that would/can be the case for Probit, or GLM with non-canonical link)

I don't know if all tsa models have mean(resid)=0 by construction/definition.

**Comment by josef-pkt** [comment]
out of curiosity I skimmed parts of the original Box-Pierce and Ljung-Box articles.
AFAIU based on the superficial skimming
Box-Pierce only talk about using chisquare distribution as an approximation. If the ARIMA parameters were known instead of estimated, then the test statistic would be (asymptotically) chisquare distributed. But they don't have the distribution when the parameters are estimated.

So, ljung-box uses only an approximate but not the correct asymptotic distribution.

Breusch-Godfrey takes the correlation/dependence of residuals on exog into account.

Box-Pierce use the argument that an AR can approximate the predictions of an ARMA, to argue that the same as derived for AR also applies to ARMA.

In analogy, we could use a larger number of lagged endog in the auxiliary regression to take account of the ARMA estimation. Then BG might still be better than BP/LB.

(I didn't look at any more recent literature.)

Back to this PR:
BG applies if we have only a constant in OLS, and BG would apply without any original exog (no constant) if mean of data is zero.
So we can make exog optional, even if it's not completely appropriate or is only approximate if we use it after ARMA (and don't include lagged endog)

**Comment by ricardozago** [comment]
@josef-pkt thank you for your replies. 

Can this test be used for ARMA models with constant?

> BG applies if we have only a constant in OLS, and BG would apply without any original exog (no constant) if mean of data is zero.
> So we can make exog optional, even if it's not completely appropriate or is only approximate if we use it after ARMA (and don't include lagged endog)

I noticed that in Statsmodels implementation, don't matter if the model have or not a constant, the test always add a constant in this part of code.

`xdall = np.c_[np.ones((nobs, 1)), xdall]`

If the model already has a constant, the coefficient of the variable will be divided by two, but the BG test result will be not affected. Do you know if is this expected?

**Comment by josef-pkt** [comment]
> If the model already has a constant, the coefficient of the variable will be divided by two, but the BG test result will be not affected. Do you know if is this expected?

The LM test is not affected by the presence of perfectly collinear variables, i.e. variables can be duplicate. This is because the rsquared version of the LM test uses rsquared which does not depend on identified params. The df for the chi2 pvalues needs to be correct, in OLS we use the rank, in this case it is fixed at nlags.

The f-test version could potentially be incorrect if the added variables are collinear with existing variables. AFAICS, we are only testing the added lags which should not be collinear with the existing exog. 
(except in corner cases, e.g. perfect fit and all resid are zero. I never saw weird cases for this test.)

**Comment by josef-pkt** [comment]
I think we can merge this as a "programming" solution, and leave usage up to the user.
It's a simple change.

**Comment by ricardozago** [comment]
> I think we can merge this as a "programming" solution, and leave usage up to the user.
> It's a simple change.

Thank you, I agree that it's a "programming" solution. Do I need to do anything for this aprove this pull request?

**Comment by josef-pkt** [comment]
I got distracted again.

Can you rebase it on current master? then we merge when green
(otherwise I will rebase it myself and merge soon)

File: statsmodels/stats/tests/test_diagnostic.py [hunk 1]
```diff
@@ -287,6 +288,11 @@ def test_acorr_breusch_godfrey_multidim(self):
         with pytest.raises(ValueError, match='Model resid must be a 1d array'):
             smsdia.acorr_breusch_godfrey(res)
 
+    def test_acorr_breusch_godfrey_exogs(self):
+        data = sunspots.load_pandas().data['SUNACTIVITY']
+        res = ARMA(data, (1,0)).fit(disp=False, trend = 'nc')
+        smsdia.acorr_breusch_godfrey(res, nlags = 1)
```
Comment (josef-pkt) [inline statsmodels/stats/tests/test_diagnostic.py:294]: pep- 8 errors, no space around `=` on keyword arguments
previous line space after `,`

Comment (ricardozago) [reply statsmodels/stats/tests/test_diagnostic.py:294]: I tried to fix in a new the syntax issues in a new commit in this branch. May you check?

**Comment by ricardozago** [comment]
@josef-pkt I remade the commit fixing the pep8 errors. Can you check if it's ok?

It's my first commit in a open-source project.

**Comment by josef-pkt** [comment]
looks good to me, merging as is.

Thanks for the PR and cleaning it up. Sorry for the delay. It was on my todo list, but I'm still stuck in another topic.
